### ISSUE TITLE
[Loopring 3.* Bounty Idea] Payment receipts (challage and response game)

### ISSUE THREAD

**Issue by Brechtpd** [Loopring/protocols#1279]
Receipts are given to a user by the operator that does a transfer. The receipt provides him some extra guarantees that the payment will actually go through and the transfer won't just be thrown away. If the transfer isn't included in a block the receipt can be used to still get the transferred amount directly from an insurance contract provided by the operator. 

This provides better finality, but because the operator can create an infinite number of receipts where he promises to include transfers in a block, but the insurance contract only has a limited amount of funds, it's certainly not a 100% guarantee the transfer will actually be done. But still pretty good.

This functionality is provided by zkSync and Starkware, and people are talking about this. If want to make an actual payment platform it's a good thing to have.

**Comment by dong77** [comment]
How receipt is used/verified/guaranteed in the smart contract?

**Comment by Brechtpd** [comment]
That is still an open question as I haven't really looked into this at all. :)

But I guess it could work similarly to light clients for blockchains. 
- For each block a hash of all transactions is saved onchain (either a flat list or in a Merkle tree format).
- The receipt the operator gives to the payer/payee is the block and position the transfer will be done in. 
- The operator submits the block. If the transfer didn't happen at the promised position in the block the payer/payee can submit a fraud proof on the insurance contract showing that a different transfer happened at that position. He can do that by posting all the DA again (or if it's in a Merkle tree format, a Merkle proof) and shows that the receipt contained a different transfer than was actually done.
- If indeed the operator lied, the receipt can be used to still transfer the funds to the payee using funds inside the insurance contract the operator created to provide instant economic finality.

So basically just by saving the data availability hash for each block onchain would already allow us to provide such functionality. We actually do that in 3.1 but don't do that any more in 3.5. 

... so I guess we can actually do this already in 3.1 if we want to! The data availability hash for each block can be queried with `getBlock`. Heh, I actually wasn't aware of this before typing this up haha. :)

But I guess there could be better ways to do it. One downside in 3.1 is that the transfer doesn't contain a nonce to make sure the transfer is unique, but that's not the biggest problem as the recipient of the receipt can just check if the receipt is actually "good" (not cheating by pointing to an old transfer or something like that) before accepting it.

**Comment by dong77** [comment]
So if we append the hash of each block in a list in 3.6, it is still hard to accept a receipt as a challenge because the data (DA) for each block is not structured (queryable), but as call data, isn't it?

Maybe we can hash each block's data into its own Merkel tree and roll up the tree root. Then they can use Merkel root to verify if a receipt is inside a given block without using the block's DA directly.

**Comment by Brechtpd** [comment]
I actually think a flat list is fine. If someone wants to submit a receipt onchain they just need to fetch the DA of the block the receipt promises the transfer was done in, submit that together with the receipt, and the DA is structured enough that we can jump efficiently to the correct location (or even check all transfers in the block if the operator didn't promise to do the transfer at a specific location in a block, but just a certain block).

I don't think it's important that submitting these receipts as fraud proof is very efficient because it shouldn't happen to begin with (and the insurance contract can repay all gas costs to the submitter of the receipt). So best to just keep doing what's most efficient in normal scenario's.

With the Merkle tree approach there is less data that needs to be submitted onchain, but it will probably end up being more expensive. To be efficient we'd have to hash the Merkle tree with Poseidon, which would then be very expensive onchain to verify so this will probably end up using more gas. The Merkle tree approach is also a little flexible because it only works efficiently if we know the exact location the transaction was done in a block. If we just use the flat list the receipt can provide some more flexible criteria (like just somewhere in a block, or some range).

So I don't know how other project do it, but it seems like we don't really need to do any changes and make things more difficult to support this in a decent way. We just need to add back saving the hash of the DA data for each block in 3.5.

**Comment by dong77** [comment]
Even if a user submits a receipt with the corresponding DA, how can our smart contract know the DA he submitted is the right one. We don't have a way for the smart contract to find the DA of a certain block onchain for verification. Or are you saying that for each block, we still need to submit the hash of the DA together with the state root? The hash is a simple hash, not a Merkel tree hash of the DA.

**Comment by Brechtpd** [comment]
Any smart contract can get the DA hash (called `publicDataHash` here) for a block in 3.1 by calling `getBlock`: https://github.com/Loopring/protocols/blob/a7a8b3e2b4b6f07405389ca2f5197ae91b48d733/packages/loopring_v3/contracts/iface/IExchangeV3.sol#L454

So by sending the complete DA again
- we have access to all transactions in the block to verify if the transfer is there
- we can hash the data again just like it was hashed in the exchange smart contract and verify the hash is the same one as returned by `getBlock`.

**Comment by dong77** [comment]
## A better receipt validation based on game theory

For an transfer `T = {from, to, amount, token, id, ...}`, the operator generates and signs a receipt:

```
R = {timestamp, T, operator, opratorSig(timestamp, hash(T))}
``` 

and returns it to the user.

The user can submit a receipt challenge on-chain before `timestamp`, but with a deposit of 0.1ETH, for example.
```
challangeReceipt(Receipt r) {...}

```

Once a receipt challenge is put onchain, the operator will have one or two weeks to respond. If within this period, the operator fails to provide information that can be used to locate T's data in one of the blocks, he will be punished or pay back the user with another fund. Otherwise, the challenge will be cleared and the 0.1ETH challenge fee will be sent to the relayer.

```
clearReceiptChallange(bytes32 receiptHash, bytes calldata, unit blockIndex, unit offSet)
```

**Comment by Brechtpd** [comment]
I think there may be a problem because the operator at any time can create invalid receipts. So if he doesn't actually do a transfer for some user, the only thing he needs to do is create a false receipt that transfers funds to himself and drain the insurance contract before any user is able to do so. I don't think we can detect these invalid receipts.
- Limiting the amount that can be redeemed per receipt doesn't help because the operator can create an infinite number of invalid receipts.
- Limiting how much can be withdrawn in certain amount of time could help, but then it's just a race who gets in first, so that doesn't really work either.

This will be likely be very obvious and damage the reputation of the operator which may be good enough to stop this though. And there is no guarantee there are enough funds in the insurance contract any way.

**Comment by Brechtpd** [comment]
I think it may be useful to also expose the timestamp the block was submitted onchain. This way we can allow receipts to specify a timestamp a transaction should have been included in a block before. Otherwise it seems with the method above the receipt could be posted onchain but the operator could include it in a block only when challenged, which wouldn't work well because posting the receipt onchian costs money (+ extra deposit) and the operator will be rewarded for this bad behaviour by doing transactions only when the receipt is posted onchain.

So with this, the recipient can be 100% the operator won't be able to still do the transaction once the "at least before" time has passed and the user posted the receipt onchain, because any newer block the operator may submit with the transaction will have a newer timestamp and so won't be usable to clear the challenge.

**Comment by dong77** [comment]
I like this idea.
